### ISSUE TITLE
Support XMB

### ISSUE THREAD

**Issue by ebursztein** [tensorflow/similarity#92]
https://github.com/bnu-wangxun/Deep_Metric

**Comment by chjort** [comment]
I would like to contribute to this issue.

One way to implement the Cross-batch-memory (XBM) module is to extend `MetricLoss` with
a memory data structure and necessary enqueue and dequeue functionality. It is also
important to enable a warmup period where XBM is not active.

In the following implementation example, the memory data structure consists of a `tf.Variable`
with dynamic size in the 0th dimension, while enqueue and dequeue are implemented with
`tf.concat` and slicing respectively.
```python
from typing import Callable, Optional, Union

import tensorflow as tf

from tensorflow_similarity.distances import Distance, distance_canonicalizer
from tensorflow_similarity.losses import MetricLoss
from tensorflow_similarity.losses.multisim_loss import multisimilarity_loss
from tensorflow_similarity.types import FloatTensor


def _add_memory_variable(tensor):
    """Creates an empty variable with same shape and dtype as `tensor`."""
    shape_no_batch = tensor.shape[1:]
    dtype = tensor.dtype
    init_value = tf.constant([], shape=[0, *shape_no_batch], dtype=dtype)
    var = tf.Variable(
        initial_value=init_value,
        shape=[None, *shape_no_batch],
        dtype=dtype,
        trainable=False,
    )
    return var


class XBM(MetricLoss):
    def __init__(
        self,
        fn: Callable,
        memory_size: int,
        warmup_steps: int = 0,
        reduction: Callable = tf.keras.losses.Reduction.AUTO,
        name: Optional[str] = None,
        **kwargs
    ):
        super().__init__(fn, reduction, name, **kwargs)

        if isinstance(fn, MetricLoss):
            self.distance = fn.distance

        self.memory_size = memory_size
        self.warmup_steps = warmup_steps
        self._y_true_memory = None
        self._y_pred_memory = None
        self._total_steps = tf.Variable(0, dtype=tf.int64, trainable=False)

    def call(self, y_true: FloatTensor, y_pred: FloatTensor) -> float:
        if self._y_true_memory is None:
            self._y_true_memory = _add_memory_variable(y_true)
        if self._y_pred_memory is None:
            self._y_pred_memory = _add_memory_variable(y_pred)

        # enqueue (concat batch to beginning of memory)
        y_true_mem = tf.concat([y_true, self._y_true_memory], axis=0)
        y_pred_mem = tf.concat([y_pred, self._y_pred_memory], axis=0)

        # dequeue (truncate to memory size limit)
        y_true_mem = y_true_mem[: self.memory_size]
        y_pred_mem = y_pred_mem[: self.memory_size]

        def _xbm_step():
            # update memory
            self._y_true_memory.assign(y_true_mem)
            self._y_pred_memory.assign(y_pred_mem)
            return y_true_mem, y_pred_mem

        def _warmup_step():
            return y_true_mem, y_pred_mem

        y_true_mem, y_pred_mem = tf.cond(
            self._total_steps > self.warmup_steps,
            true_fn=_xbm_step,
            false_fn=_warmup_step,
        )
        self._total_steps.assign_add(1)

        return super().call(y_true_mem, y_pred_mem)
```

This implementation can be used either as a wrapper around existing `MetricLoss` loss instances:
```python
loss = MultiSimilarityLoss(distance="cosine")
loss = XBM(loss, memory_size=100, warmup_steps=1000)
model.compile("adam", loss=loss)
```

or as a parent class to inherit XBM functionality:
```python
class XBMMultiSimilarityLoss(XBM):
    def __init__(
        self,
        memory_size: int,
        distance: Union[Distance, str] = "cosine",
        alpha: float = 1.0,
        beta: float = 20,
        epsilon: float = 0.2,
        lmda: float = 0.5,
        warmup_steps: int = 0,
        name: str = "XBMMultiSimilarityLoss",
    ):
        distance = distance_canonicalizer(distance)
        self.distance = distance

        super().__init__(
            multisimilarity_loss,
            name=name,
            reduction=tf.keras.losses.Reduction.NONE,
            distance=distance,
            alpha=alpha,
            beta=beta,
            epsilon=epsilon,
            lmda=lmda,
            memory_size=memory_size,
            warmup_steps=warmup_steps,
        )


loss = XBMMultiSimilarityLoss(distance="cosine", memory_size=100, warmup_steps=1000)
model.compile("adam", loss=loss)
```

This trains just fine
```
Epoch 1/5
1000/1000 [==============================] - 18s 16ms/step - loss: 0.8051  # warmup steps
Epoch 2/5
1000/1000 [==============================] - 16s 16ms/step - loss: 1.8837
Epoch 3/5
1000/1000 [==============================] - 16s 16ms/step - loss: 1.7174
Epoch 4/5
1000/1000 [==============================] - 16s 16ms/step - loss: 1.5286
Epoch 5/5
1000/1000 [==============================] - 16s 16ms/step - loss: 1.3939
```

However, there is one problem with the current `Distance` class in order to correctly support XBM.
Right now the `call` method of every `Distance` has the following signature:

```python
def call(self, embeddings: FloatTensor) -> FloatTensor:
```

This only allows computing the distances between the embeddings themselves like _d(embeddings, embeddings)_,
where _d_ is some mathematical distance function. Therefore the current XBM implementation can only compute
the distances between the memory embeddings themselves like so _d(memory_embeddings, memory_embeddings)_.

The XBM paper you referenced requires the distances to be computed between batch embeddings and memory embeddings like so
_d(batch_embeddings, memory_embeddings)_. So a refactoring of the `Distance` class might be necessary in order to
properly support XBM.

We have to consider if the `call` method of `Distance` classes should support computing distances between `embeddings`
and some optional `target_embeddings` like so:

```python
def call(self, embeddings: FloatTensor, target_embeddings: Optional[FloatTensor] = None) -> FloatTensor:
    if target_embeddings is not None:
        return d(embeddings, target_embeddings)
    else:
        return d(embeddings, embeddings)
```

or perhaps like this:
```python
def call(self, embeddings: Union[FloatTensor, Tuple[FloatTensor, FloatTensor]]) -> FloatTensor:
    if isinstance(embeddings, tuple):
        embeddings, target_embeddings = embeddings
        return d(embeddings, target_embeddings)
    else:
        return d(embeddings, embeddings)
```

Please let me know what you think!

**Comment by owenvallis** [comment]
Hi @chjort, thanks for working on this! I like the implementation, especially how this supports using the existing MetricLoss classes.

Regarding the Distance signature, I think I prefer adding a second parameter as an Optional[FlaotTensor] over the more complex Union of types. Alternatively, we could explicitly require both embedding inputs. This would be more verbose, but would avoid the conditional and make it clearer which two embeddings we are comparing. One downside is this would be a breaking change.

```
def call(self, source_embeddings: FloatTensor, target_embeddings: FloatTensor) -> FloatTensor:
     ...
```

@ebursztein wdyt?

**Comment by ebursztein** [comment]
I think it's a reasonable change if we document it carefully. I would be more comfortable if we also think through how this will work with prototype network e.g SWAV to make sure we only change the signature once.

Make sense?

**Comment by chjort** [comment]
Comparison between two different sets of embeddings is central in XBM, but also in a multitude of self-supervised contrastive approaches including:

* SvAV
* Barlow Twins
* SimCLR
* BYOL
* SimSiam

Many of these self-supervised approaches have specialized loss functions which do not expect any lables, but instead computes a loss from the embeddings only (often by comparing embeddings from different views (augmentations) of a batch).

Changing `Distance` to require two embeddings will allow these self-supervised losses to also utilize the `Distance` class, and not having to re-implement the distance function within the self-supervised loss. Essentially expanding the set of use-cases `Distance` can support.

I will start working on a Pull Request where I refactor the `Distance` class to have the following call signature:
```python
def call(self, query_embeddings: FloatTensor, key_embeddings: FloatTensor) -> FloatTensor:
```
and make the necessary changes up stream to support all current functionality of the __master__ branch (mostly the _losses_ module is affected here). If changes to the `Distance` class gets merged, it will make sense to continue working on XBM.

**Comment by chjort** [comment]
The refactoring of distances and losses is now proposed in PR #222, making it easier to implement XBM in a new future pull request if #222 gets merged.
